File: semester_4/numerical_methods/one_dimensional_optimization/golden_ratio.py
from collections.abc import Callable


def golden_ratio_search(f: Callable, a: float, b: float, eps: float) -> float:
    tau = 0.618

    while abs(a - b) < eps:
        x_1 = b - ((b - a) / tau)
        x_2 = a + ((b - a) / tau)
        if f(x_1) >= f(x_2):
            a = x_1
        else:
            b = x_2

    return f((a + b) / 2)

File: semester_4/optimization/lw1_one_dimensional_optimization/src/golden_ratio.rs
pub fn golden_ratio_search<F>(f: F, a: f64, b: f64, eps: f64) -> (f64, f64)
where
    F: Fn(f64) -> f64,
{
    let mut a = a;
    let mut b = b;

    let tau = 0.618;

    let mut x_1 = b - (b - a) * tau;
    let mut x_2 = a + (b - a) * tau;

    while (a - b).abs() > eps {
        if f(x_1) < f(x_2) {
            b = x_2;
        } else {
            a = x_1;
        }

        x_1 = b - (b - a) * tau;
        x_2 = a + (b - a) * tau;
    }

    ((a + b) / 2.0, f((a + b) / 2.0))
}
